src/MetaMorpho.sol

```diff
@@ -345,51 +345,52 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
+            (uint256 totalSupplyAssets, uint256 totalSupplyShares,,) =
+                MORPHO.expectedMarketBalances(allocation.marketParams);
 
-                allocation.assets = 0;
-            }
+            uint256 supplyShares = MORPHO.supplyShares(id, address(this));
+            uint256 supplyAssets = supplyShares.toAssetsDown(totalSupplyAssets, totalSupplyShares);
+            uint256 withdrawn = supplyAssets.zeroFloorSub(allocation.assets);
 
-            (uint256 withdrawnAssets, uint256 withdrawnShares) =
-                MORPHO.withdraw(allocation.marketParams, allocation.assets, shares, address(this), address(this));
+            if (withdrawn > 0) {
+                // Guarantees that unknown frontrunning donations can be withdrawn, in order to disable a market.
+                uint256 shares;
+                if (allocation.assets == 0) {
+                    shares = supplyShares;
+                    withdrawn = 0;
+                }
 
-            totalWithdrawn += withdrawnAssets;
+                (uint256 withdrawnAssets, uint256 withdrawnShares) =
+                    MORPHO.withdraw(allocation.marketParams, withdrawn, shares, address(this), address(this));
 
-            emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
-        }
+                emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
 
-        uint256 totalSupplied;
-        for (uint256 i; i < supplied.length; ++i) {
-            MarketAllocation memory allocation = supplied[i];
-            Id id = allocation.marketParams.id();
-            uint256 supplyCap = config[id].cap;
+                totalWithdrawn += withdrawnAssets;
+            } else {
+                uint256 supplied = allocation.assets.zeroFloorSub(supplyAssets);
+                if (supplied == 0) continue;
 
-            if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
+                uint256 supplyCap = config[id].cap;
+                if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
 
-            (uint256 suppliedAssets, uint256 suppliedShares) =
-                MORPHO.supply(allocation.marketParams, allocation.assets, 0, address(this), hex"");
+                if (supplyAssets + supplied > supplyCap) revert ErrorsLib.SupplyCapExceeded(id);
 
-            if (_supplyBalance(allocation.marketParams) > supplyCap) {
-                revert ErrorsLib.SupplyCapExceeded(id);
-            }
+                (uint256 suppliedAssets, uint256 suppliedShares) =
+                    MORPHO.supply(allocation.marketParams, supplied, 0, address(this), hex"");
 
-            totalSupplied += suppliedAssets;
+                emit EventsLib.ReallocateSupply(_msgSender(), id, suppliedAssets, suppliedShares);
 
-            emit EventsLib.ReallocateSupply(_msgSender(), id, suppliedAssets, suppliedShares);
+                totalSupplied += suppliedAssets;
```

Comment:
Curious to know why you switched those 2 lines :eyes:

Reply:
This is detail, but I feel like:
- emitting the event right after the CALL is cleaner than after having increased the memory variable
- it allows to free the stack from `suppliedShares`